encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+	if err != nil {
+		return err
+	}
+
+	if len(parameterTypes) == 1 {
+		// Avoid overhead of sorting if there is only one parameter type.
+		return e.encodeParameterTypeValue(parameterTypes[0], visited)
+	}
+
+	// "Deterministic CCF Encoding Requirements" in CCF specs:
+	//
+	//   "composite-type-value.initializers MUST be sorted by identifier."
+
+	// NOTE: bytewiseParmaeterSorter doesn't sort parameterTypes in place.
+	// bytewiseParmaeterSorter.indexes is used as sorted parameterTypes
+	// index.
+	sorter := newBytewiseParameterSorter(parameterTypes)
+
+	sort.Sort(sorter)
+
+	// Encode sorted parameter types.
+	for i := 0; i < len(sorter.indexes); i++ {
+		index := sorter.indexes[i]
+
+		err = e.encodeParameterTypeValue(parameterTypes[index], visited)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// encodeParameterTypeValue encodes composite initializer parameter as
+// language=CDDL
+//
+//	 [
+//	    label: tstr,
+//	    identifier: tstr,
+//	    type: type-value
+//	]
+func (e *Encoder) encodeParameterTypeValue(parameterType cadence.Parameter, visited ccfTypeIDByCadenceType) error {
+	// Encode CBOR array head with length 3
+	err := e.enc.EncodeRawBytes([]byte{
+		// array, 3 items follow
+		0x83,
+	})
+	if err != nil {
+		return err
+	}
+
+	// element 0: label as tstr.
+	err = e.enc.EncodeString(parameterType.Label)
+	if err != nil {
+		return err
+	}
+
+	// element 1: identifier as tstr.
+	err = e.enc.EncodeString(parameterType.Identifier)
+	if err != nil {
+		return err
+	}
+
+	// element 2: type as type-value.
+	return e.encodeTypeValue(parameterType.Type, visited)
+}
+
+// encodeFunctionTypeValue encodes cadence.FunctionType as
+// language=CDDL
+// function-type-value =
+//
+//	; cbor-tag-function-type-value
+//	#6.193(function-value)
+func (e *Encoder) encodeFunctionTypeValue(typ *cadence.FunctionType, visited ccfTypeIDByCadenceType) error {
+	// Encode tag number and array head of length 3.
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, CBORTagFunctionTypeValue,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Encode function-value.
+	return e.encodeFunction(typ, visited)
+}
+
+// encodeNilTypeValue encodes nil type value as CBOR nil.
+func (e *Encoder) encodeNilTypeValue() error {
+	return e.enc.EncodeNil()
+}
+
+// needToEncodeType returns true if runtimeType needs to be encoded because:
+// - static type is missing (top level value doesn't have static type)
+// - static type is different from runtime type (static type is abstract type)
+func needToEncodeRuntimeType(staticType cadence.Type, runtimeType cadence.Type) bool {
+	if staticType == nil {
+		return true
+	}
+	if staticType.Equal(runtimeType) {
+		return false
+	}
+	// Here, static type is different from runtime type.
+	// Handle special case of runtime type being OptionalType{NeverType}.
+	if _, ok := staticType.(*cadence.OptionalType); ok {
+		if isOptionalNeverType(runtimeType) {
+			return false
+		}
+	}
+	return true
```

Comment:
Not sure, but why do we have to handle `Never?` specially?

Reply:
We handle special case of `Optional{nil}` because its runtime type is `OptionalType{NeverType}` while its static type can be different, such as `OptionalType{AddressType}`.

In this [test example](https://github.com/onflow/cadence/blob/705fdd57ffb42d9f68fe3904e2fbc0aace362751/encoding/ccf/ccf_test.go#L7925-L7939), field `to`'s type is `OptionalType{AddressType}` and its value is nil with runtime type `OptionalType{NeverType}`.  Even though runtime type is different from static type (field type), encoder encodes `nil` value without encoding its runtime type.

Reply:
Added comment in commit 49d23b175ccf510f635c9e8fe8cdc68eefbb649a.